[16.0][MIG] attachment_delete_restrict: Migration to 16.0

AungKoKoLin1997 opened the pull request:
@qrtl

This PR encompasses the following steps:

1. **Standard Migration**
2. **Fixes**
        - Removal of the display of duplicate allowed user names.
        - Correct the many2many field relation 
        - Correction of domain values and addition of the many2many_tags widget in views.
3. **Improvements**
        - Addition of _onchange_restrict_delete_attachment() to clear groups and users when restrict_delete_attachment is not set to custom or owner_custom.

yostashiro committed:
[ADD] attachment_delete_restrict

Kev-Roche committed:
[IMP] attachment_delete_restrict: black, isort, prettier

Kev-Roche committed:
[MIG] attachment_delete_restrict: Migration to  14.0

Kev-Roche committed:
rework module attachment_delete_restrict (to squash after review)

sebastienbeau committed:
attachment_delete_restrict: refactor test and improve restrict mode naming

sebastienbeau committed:
attachment_delete_restrict: refactor code

Commit:
[UPD] Update attachment_delete_restrict.pot

OCA-git-bot committed:
[UPD] README.rst

OCA-git-bot committed:
[ADD] icon.png

ibuioli committed:
Added translation using Weblate (Spanish (Argentina))

ibuioli committed:
Translated using Weblate (Spanish (Argentina))

Currently translated at 100.0% (37 of 37 strings)

Translation: server-tools-14.0/server-tools-14.0-attachment_delete_restrict
Translate-URL: https://translation.odoo-community.org/projects/server-tools-14-0/server-tools-14-0-attachment_delete_restrict/es_AR/

Commit:
[FIX] attachment_delete_restrict: adapt to Odoo permission change

OCA-git-bot committed:
attachment_delete_restrict 14.0.1.0.1

AungKoKoLin1997 committed:
[MIG] attachment_delete_restrict: Migration to 16.0

yostashiro commented on attachment_delete_restrict/models/ir_attachment.py:
```diff
@@ -33,7 +33,7 @@ def _raise_delete_attachment_error(self, allowed_users):
                 "the delete permission:\n%s"
             )
-            % ("\n".join(allowed_users.mapped("name")) or "None")
+            % ("\n".join(list(set(allowed_users.mapped("name")))) or "None")
```
This `set()` should be done when `allowed_users` is updated in `_check_custom_delete_attachment()` instead.  It should perhaps be in a separate commit to be backported.

yostashiro commented on attachment_delete_restrict/models/ir_model.py:
```diff
@@ -0,0 +1,50 @@
+    )
+
+    @api.onchange("restrict_delete_attachment")
+    def _clear_groups_users(self):
```
```suggestion
    def _onchange_restrict_delete_attachment(self):
```

I suppose this should also be in a separate commit to be backported.

yostashiro commented on attachment_delete_restrict/models/res_groups.py:
```diff
@@ -0,0 +1,29 @@
+        string="Attachment Deletion Models",
+        help="The users of the group can delete the attachments related to "
+        "the models "
+        "assigned here.",
```
```suggestion
        "the models assigned here. Restrict Attachment Deletion must be 'Custom' or 'Owner + Custom' for the model to be selected here.",
```

yostashiro commented on attachment_delete_restrict/models/res_groups.py:
```diff
@@ -0,0 +1,29 @@
+                                "delete_attachment_group_ids": [(4, group.id)],
+                            }
+                        )
+        return res
```
Can you explain why this should be needed?

yostashiro commented on attachment_delete_restrict/models/res_users.py:
```diff
@@ -0,0 +1,28 @@
+                                "delete_attachment_user_ids": [(4, user.id)],
+                            }
+                        )
+        return res
```
Why do we need this?

yostashiro reviewed (changes requested):
@AungKoKoLin1997 I think some of the changes should be in separate commits to be backported to 14.0.

AungKoKoLin1997 committed:
[FIX] attachment_delete_restrict
- Remove the display of duplicate allowed user names.
- Correct the many2many field relation
- Correct domain values and add the many2many_tags widget in views.

AungKoKoLin1997 committed:
[IMP] attachment_delete_restrict
- Add _onchange_restrict_delete_attachment() to clear groups and users when restrict_delete_attachment is not set to custom or owner_custom.

yostashiro commented on attachment_delete_restrict/models/ir_attachment.py:
```diff
@@ -59,14 +59,14 @@ def _check_custom_delete_attachment(self, model=None, allow_owner_and_admin=Fals
         res_models = list(set(self.filtered("res_model").mapped("res_model")))
         if res_models:
-            models = self.env["ir.model"].search([("model", "in", res_models)])
+            models = self.env["ir.model"].sudo().search([("model", "in", res_models)])
```
This should be part of the migration, due to the ACL change in 16.0 (or 15.0, I haven't checked).

legalsylvain commented:
hi @yostashiro. Thanks for your review. could you update it ? 
thanks !

AungKoKoLin1997 commented:
@OCA-server-tools-maintainers
Is it possible to review this PR?

yostashiro reviewed (approved):
Code review and functional test. LGTM.

kanda999 reviewed (approved):
Code and Function : LGTM

kanda999 commented:
@AungKoKoLin1997 
Could you rebase this PR to restart runboat?

AungKoKoLin1997 commented:
@legalsylvain Can you please review?

sebastienbeau committed:
attachment_delete_restrict: using sudo should by pass the restriction

hitrosol commented:
can we have the runboat available ?

yostashiro commented:
@hitrosol runboat is up now. ;)

hitrosol commented:
Functionally test LGTM.
Having date for optional criteria will be better as some case we don't allow common users delete attachment before certain date but it would be okay after certain date, but i think it can done in the next PR.

yostashiro reviewed (commented):
@AungKoKoLin1997 Can you please cherry-pick the changes from https://github.com/OCA/server-tools/pull/3248.

AungKoKoLin1997 commented:
> Can you please cherry-pick the changes from #3248.

@yostashiro Done!

yostashiro commented:
/ocabot merge nobump
